File: VB/DevExpress.Example03/CheckableGroupBehavior.vb
' Developer Express Code Central Example:
' How to create Checkable Grouping in GridControl
' 
' The example demonstrates how to add the CheckBox functionality to each GroupRow
' in TableView. The basic idea is to allow a user to easily check or uncheck the
' necessary group of items in GridControl.
' 
' The functionality is realized as
' behavior in the CheckableGroupBehavior class, which can be attached to
' GridControl. It automatically sets GroupValueTemplate using the GroupCheckBox
' class, which is inherited from the CheckBox class. The CheckableGroupBehavior's
' CheckableProperty must be set and has to contain the name of the property in a
' row data object, which will be used to check items. The property has to be of
' the Boolean type.
' 
' You can easily add the same functionality to your project by
' using the CheckableGroupBehavior class and attaching it as behavior to your
' GridControl.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=T127563

Imports DevExpress.Xpf.Grid
Imports DevExpress.Mvvm.UI.Interactivity
Imports System
Imports System.Collections.Generic
Imports System.Collections.Specialized
Imports System.ComponentModel
Imports System.Linq
Imports System.Windows
Imports System.Windows.Markup

Namespace DevExpress.Example03

    Public Class CheckableGroupBehavior
        Inherits Behavior(Of GridControl)

        #Region "Fields"

        Protected _LocalSet As Boolean

        #End Region ' Fields

        #Region "Checkable"

        Public Property CheckableProperty() As String
            Get
                Return CStr(GetValue(CheckablePropertyProperty))
            End Get
            Set(ByVal value As String)
                SetValue(CheckablePropertyProperty, value)
            End Set
        End Property

        Public Shared ReadOnly CheckablePropertyProperty As DependencyProperty = DependencyProperty.Register("CheckableProperty", GetType(String), GetType(CheckableGroupBehavior), New PropertyMetadata(String.Empty))

        #End Region ' Checkable

        #Region "CheckList"

        Public Shared Function GetCheckList(ByVal obj As DependencyObject) As Dictionary(Of Integer, Boolean?)
            Return CType(obj.GetValue(CheckListProperty), Dictionary(Of Integer, Boolean?))
        End Function

        Public Shared Sub SetCheckList(ByVal obj As DependencyObject, ByVal value As Dictionary(Of Integer, Boolean?))
            obj.SetValue(CheckListProperty, value)
        End Sub

        Public Shared ReadOnly CheckListProperty As DependencyProperty = DependencyProperty.RegisterAttached("CheckList", GetType(Dictionary(Of Integer, Boolean?)), GetType(CheckableGroupBehavior), New PropertyMetadata(New Dictionary(Of Integer, Boolean?)(), AddressOf IsCheckListPropertyChanged))

        #End Region ' CheckList

        Public Shared Event IsCheckListChanged As Action(Of DependencyObject, DependencyPropertyChangedEventArgs)

        Protected Shared Sub IsCheckListPropertyChanged(ByVal obj As DependencyObject, ByVal args As DependencyPropertyChangedEventArgs)
            RaiseEvent IsCheckListChanged(obj, args)
        End Sub

        Protected Sub OnIsCheckListChanged(ByVal sender As Object, ByVal args As DependencyPropertyChangedEventArgs)
            If args.NewValue Is Nothing Then
                Return
            End If

            If Me._LocalSet Then
                Return
            End If

            Me._LocalSet = True
            Dim grid = Me.AssociatedObject
            Dim table = TryCast(Me.AssociatedObject.View, TableView)
            Dim rowHandle = grid.GetSelectedRowHandles().First()
            Dim checkStates = (TryCast(args.NewValue, Dictionary)(Of Integer, Boolean?))
            Dim groupLevel As Integer = grid.GetRowLevelByRowHandle(rowHandle)
            Dim value? As Boolean = If(checkStates.Keys.Contains(groupLevel), checkStates(groupLevel), False)
            Me.WorkOutGroup(rowHandle, value)
            Me._LocalSet = False
        End Sub

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            If Me.AssociatedType IsNot GetType(GridControl) OrElse Me.CheckableProperty = String.Empty Then
                Return
            End If

            AddHandler Me.AssociatedObject.Loaded, AddressOf GridLoaded
        End Sub

        Protected Sub AttachItem(ByVal obj As Object)
            Dim iNotifyPropertyChanged = DirectCast(obj, INotifyPropertyChanged)
            If iNotifyPropertyChanged IsNot Nothing Then
                AddHandler iNotifyPropertyChanged.PropertyChanged, AddressOf Me.PropertyChanged
            End If
        End Sub

        Protected Sub GridLoaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
            If Me.AssociatedObject.View.GetType() Is GetType(TableView) Then
                TryCast(Me.AssociatedObject.View, TableView).GroupValueTemplate = Me.GetGroupValueTemplate()
            End If

            If Me.AssociatedObject.GroupCount > 0 Then
                Me.AssociatedObject_EndGrouping(Nothing, Nothing)
            End If

            Dim items = TryCast(Me.AssociatedObject.ItemsSource, IEnumerable(Of Object))
            If items IsNot Nothing Then
                For Each item In items
                    Me.AttachItem(item)
                Next item
            End If

            Dim collection As INotifyCollectionChanged = DirectCast(Me.AssociatedObject.ItemsSource, INotifyCollectionChanged)
            If collection IsNot Nothing Then
                AddHandler collection.CollectionChanged, AddressOf CollectionChanged
            End If

            AddHandler IsCheckListChanged, AddressOf OnIsCheckListChanged
            AddHandler Me.AssociatedObject.EndGrouping, AddressOf AssociatedObject_EndGrouping
            AddHandler TryCast(Me.AssociatedObject.View, TableView).CellValueChanged, AddressOf GridCellValueChanged
            AddHandler TryCast(Me.AssociatedObject.View, TableView).CellValueChanging, AddressOf CellValueChanging
        End Sub

        Private Sub AssociatedObject_EndGrouping(ByVal sender As Object, ByVal e As RoutedEventArgs)
            If Me.AssociatedObject.GroupCount = 0 Then
                Return
            End If

            Dim res = Me.GetAllRowHandles()
            Dim parents As New List(Of Integer)()
            Dim j As Integer = res.Count
            j -= 1
            Do While j >= 0
                Dim parent = Me.AssociatedObject.GetParentRowHandle(res(j))
                If Not parents.Contains(parent) Then
                    parents.Add(parent)
                    Do
                        parent = Me.AssociatedObject.GetParentRowHandle(parent)
                        If (Not parents.Contains(parent)) AndAlso Me.AssociatedObject.IsValidRowHandle(parent) Then
                            parents.Add(parent)
                        End If

                    Loop While Me.AssociatedObject.IsValidRowHandle(parent)
                End If
                j -= 1
            Loop

            Dim g As Integer = Me.AssociatedObject.GroupCount
            g -= 1
            Do While g >= 0
                Dim levelItems = parents.Where(Function(c) Me.AssociatedObject.GetRowLevelByRowHandle(c) = g)
                If levelItems Is Nothing Then
                    g -= 1
                    Continue Do
                End If
                For Each item In levelItems
                    Dim b = Me.GetRowChildrenState(item)
                    Dim dic = CType(Me.AssociatedObject.GetRowState(item).GetValue(CheckableGroupBehavior.CheckListProperty), Dictionary(Of Integer, Boolean?))
                    dic(g) = b
                    Me._LocalSet = True
                    Me.AssociatedObject.GetRowState(item).SetValue(CheckableGroupBehavior.CheckListProperty, New Dictionary(Of Integer, Boolean?)(dic))
                    Me._LocalSet = False
                Next item
                g -= 1
            Loop
        End Sub

        Protected Function GetAllRowHandles() As List(Of Integer)
            Dim result As New List(Of Integer)()
            For Each item In TryCast(Me.AssociatedObject.ItemsSource, IEnumerable(Of Object))
                Dim handle As Integer = Me.AssociatedObject.FindRowByValue("", item)
                result.Add(handle)
            Next item

            Return result
        End Function

        Protected Sub CollectionChanged(ByVal sender As Object, ByVal e As System.Collections.Specialized.NotifyCollectionChangedEventArgs)
            Select Case e.Action
                Case System.Collections.Specialized.NotifyCollectionChangedAction.Add, System.Collections.Specialized.NotifyCollectionChangedAction.Replace
                    For Each item In e.NewItems
                        Me.AttachItem(item)
                    Next item
                Case Else
            End Select
        End Sub

        Public Sub PropertyChanged(ByVal obj As Object, ByVal args As PropertyChangedEventArgs)
            If (Not Me._LocalSet) AndAlso args.PropertyName = Me.CheckableProperty Then
                Dim grid = Me.AssociatedObject
                Dim rowHandle As Integer = grid.FindRowByValue("", obj)
                Me.WorkOutCell(rowHandle)
            End If
        End Sub

        Protected Sub CellValueChanging(ByVal sender As Object, ByVal e As CellValueChangedEventArgs)
            Dim grid = Me.AssociatedObject
            grid.View.PostEditor()
        End Sub

        Protected Sub GridCellValueChanged(ByVal sender As Object, ByVal e As CellValueChangedEventArgs)
            If e.Column.FieldName = Me.CheckableProperty AndAlso (Not Me._LocalSet) Then
                Me._LocalSet = True
                Me.WorkOutCell(e.RowHandle)
                Me._LocalSet = False
            End If
        End Sub

        Protected Sub WorkOutCell(ByVal rowHandle As Integer)
            Dim grid = Me.AssociatedObject
            Dim groupRowHandle As Integer = grid.GetParentRowHandle(rowHandle)
            If grid.IsValidRowHandle(groupRowHandle) AndAlso grid.IsGroupRowHandle(groupRowHandle) Then
                Dim value? As Boolean = Me.GetRowChildrenState(groupRowHandle)
                Dim dic = New Dictionary(Of Integer, Boolean?)(Me.GetUpdatedDic(groupRowHandle, value))
                grid.GetRowState(groupRowHandle).SetCurrentValue(CheckableGroupBehavior.CheckListProperty, dic)
                Me.WorkOutCell(groupRowHandle)
            End If
        End Sub

        Protected Function GetRowChildrenState(ByVal rowHandle As Integer) As Boolean?
            Dim grid = Me.AssociatedObject
            Dim result? As Boolean = Nothing, b? As Boolean
            If grid.IsValidRowHandle(rowHandle) AndAlso grid.IsGroupRowHandle(rowHandle) Then
                Dim count As Integer = grid.GetChildRowCount(rowHandle)
                Dim i As Integer = -1
                i += 1
                Do While i < count
                    Dim childHandle = grid.GetChildRowHandle(rowHandle, i)
                    Dim childLevel As Integer = grid.GetRowLevelByRowHandle(childHandle)
                    If grid.IsGroupRowHandle(childHandle) Then
                        Dim dic = CType(grid.GetRowState(childHandle).GetValue(CheckableGroupBehavior.CheckListProperty), Dictionary(Of Integer, Boolean?))
                        b = dic(childLevel)
                    Else
                        Dim column = grid.Columns.Where(Function(c) c.FieldName = Me.CheckableProperty).FirstOrDefault()
                        If column IsNot Nothing Then
                            b = CBool(grid.GetCellValue(childHandle, column))
                        Else
                            b = False
                        End If
                    End If

                    If b.HasValue Then
                        If result.HasValue Then
                            If result.Value Xor b.Value Then
                                Return Nothing
                            End If
                        Else
                            result = b
                        End If
                    Else
                        Return Nothing
                    End If
                    i += 1
                Loop
            End If

            Return result
        End Function

        Protected Function GetUpdatedDic(ByVal rowHandle As Integer, ByVal value? As Boolean) As Dictionary(Of Integer, Boolean?)
            Dim result As Dictionary(Of Integer, Boolean?) = Nothing
            If Me.AssociatedObject.IsValidRowHandle(rowHandle) Then
                Dim groupLevel As Integer = Me.AssociatedObject.GetRowLevelByRowHandle(rowHandle)
                Dim rowState = Me.AssociatedObject.GetRowState(rowHandle)
                result = CType(rowState.GetValue(CheckableGroupBehavior.CheckListProperty), Dictionary(Of Integer, Boolean?))
                result(groupLevel) = value
            End If

            Return result
        End Function

        Protected Sub WorkOutGroup(ByVal rowHandle As Integer, ByVal value? As Boolean)
            Dim grid = Me.AssociatedObject
            Dim parentRowState = grid.GetRowState(rowHandle)
            If grid.IsValidRowHandle(rowHandle) AndAlso grid.IsGroupRowHandle(rowHandle) Then
                Dim count As Integer = grid.GetChildRowCount(rowHandle)
                Dim i As Integer = -1
                i += 1
                Do While i < count
                    Dim childHandle = grid.GetChildRowHandle(rowHandle, i)
                    If grid.IsGroupRowHandle(childHandle) Then
                        Dim rowState = Me.AssociatedObject.GetRowState(childHandle)
                        rowState.SetCurrentValue(CheckableGroupBehavior.CheckListProperty, New Dictionary(Of Integer, Boolean?)(Me.GetUpdatedDic(childHandle, value)))
                        Me.WorkOutGroup(childHandle, value)
                    Else
                        Dim column = grid.Columns.Where(Function(c) c.FieldName = Me.CheckableProperty).FirstOrDefault()
                        If column IsNot Nothing Then
                            grid.SetCellValue(childHandle, column, value)
                        End If
                    End If
                    i += 1
                Loop

                Me.WorkOutCell(rowHandle)
            End If
        End Sub

        #Region "Template For GroupRowValue"

        Protected Function GetGroupValueTemplate() As DataTemplate
            Dim xamlDataTemplate As String = "<DataTemplate>"
            xamlDataTemplate &= "<self:GroupCheckBox Content=""{Binding Value}"" Margin=""8,0,0,0""  CheckStates=""{Binding Path=RowData.RowState.(self:CheckableGroupBehavior.CheckList), Mode=TwoWay}""/>"
            xamlDataTemplate &= "</DataTemplate>"
            Dim context As New ParserContext()
            context.XmlnsDictionary.Add("", "http://schemas.microsoft.com/winfx/2006/xaml/presentation")
            context.XmlnsDictionary.Add("x", "http://schemas.microsoft.com/winfx/2006/xaml")
            context.XmlnsDictionary.Add("self", "clr-namespace:DevExpress.Example03;assembly=DevExpress.Example03")
            Dim template As DataTemplate = CType(XamlReader.Parse(xamlDataTemplate, context), DataTemplate)
            Return template
        End Function

        #End Region ' Template for GroupRowValue
    End Class
End Namespace

File: CS/DevExpress.Example03/CheckableGroupBehavior.cs
// Developer Express Code Central Example:
// How to create Checkable Grouping in GridControl
// 
// The example demonstrates how to add the CheckBox functionality to each GroupRow
// in TableView. The basic idea is to allow a user to easily check or uncheck the
// necessary group of items in GridControl.
// 
// The functionality is realized as
// behavior in the CheckableGroupBehavior class, which can be attached to
// GridControl. It automatically sets GroupValueTemplate using the GroupCheckBox
// class, which is inherited from the CheckBox class. The CheckableGroupBehavior's
// CheckableProperty must be set and has to contain the name of the property in a
// row data object, which will be used to check items. The property has to be of
// the Boolean type.
// 
// You can easily add the same functionality to your project by
// using the CheckableGroupBehavior class and attaching it as behavior to your
// GridControl.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=T127563

using DevExpress.Xpf.Grid;
using DevExpress.Mvvm.UI.Interactivity;
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Linq;
using System.Windows;
using System.Windows.Markup;

namespace DevExpress.Example03 {

    public class CheckableGroupBehavior : Behavior<GridControl> {

        #region Fields

        protected bool _LocalSet;

        #endregion Fields

        #region Checkable

        public string CheckableProperty {
            get { return (string)GetValue(CheckablePropertyProperty); }
            set { SetValue(CheckablePropertyProperty, value); }
        }

        public static readonly DependencyProperty CheckablePropertyProperty =
            DependencyProperty.Register("CheckableProperty", typeof(string), typeof(CheckableGroupBehavior), new PropertyMetadata(string.Empty));

        #endregion Checkable

        #region CheckList

        public static Dictionary<int, bool?> GetCheckList(DependencyObject obj) {
            return (Dictionary<int, bool?>)obj.GetValue(CheckListProperty);
        }

        public static void SetCheckList(DependencyObject obj, Dictionary<int, bool?> value) {
            obj.SetValue(CheckListProperty, value);
        }

        public static readonly DependencyProperty CheckListProperty =
            DependencyProperty.RegisterAttached("CheckList", typeof(Dictionary<int, bool?>), typeof(CheckableGroupBehavior), new PropertyMetadata(new Dictionary<int, bool?>(), IsCheckListPropertyChanged));

        #endregion CheckList

        public static event Action<DependencyObject, DependencyPropertyChangedEventArgs> IsCheckListChanged;
        
        protected static void IsCheckListPropertyChanged(DependencyObject obj, DependencyPropertyChangedEventArgs args) {
            if(IsCheckListChanged != null) {
                IsCheckListChanged(obj, args);
            }
        }

        protected void OnIsCheckListChanged(object sender, DependencyPropertyChangedEventArgs args) {
            if(args.NewValue == null) {
                return;
            }

            if(this._LocalSet) { return; }

            this._LocalSet = true;
            var grid = this.AssociatedObject;
            var table = this.AssociatedObject.View as TableView;
            var rowHandle = grid.GetSelectedRowHandles().First();
            var checkStates = (args.NewValue as Dictionary<int, bool?>);
            int groupLevel = grid.GetRowLevelByRowHandle(rowHandle);
            bool? value = checkStates.Keys.Contains(groupLevel) ? checkStates[groupLevel] : false;
            this.WorkOutGroup(rowHandle, value);
            this._LocalSet = false;
        }
        
        protected override void OnAttached() {
            base.OnAttached();
            if(this.AssociatedType != typeof(GridControl) || this.CheckableProperty == string.Empty) {
                return;
            }
            
            this.AssociatedObject.Loaded += GridLoaded;
        }

        protected void AttachItem(object obj) {
            var iNotifyPropertyChanged = (INotifyPropertyChanged)obj;
            if(iNotifyPropertyChanged != null) {
                iNotifyPropertyChanged.PropertyChanged += this.PropertyChanged;
            }
        }

        protected void GridLoaded(object sender, RoutedEventArgs e) {
            if(this.AssociatedObject.View.GetType() == typeof(TableView)) {
                (this.AssociatedObject.View as TableView).GroupValueTemplate = this.GetGroupValueTemplate();
            }

            if(this.AssociatedObject.GroupCount > 0) {
                this.AssociatedObject_EndGrouping(null, null);
            }

            var items = this.AssociatedObject.ItemsSource as IEnumerable<object>;
            if(items != null) {
                foreach(var item in items) {
                    this.AttachItem(item);
                }
            }

            INotifyCollectionChanged collection = (INotifyCollectionChanged)this.AssociatedObject.ItemsSource;
            if(collection != null) {
                collection.CollectionChanged += CollectionChanged;
            }

            IsCheckListChanged += OnIsCheckListChanged;
            this.AssociatedObject.EndGrouping += AssociatedObject_EndGrouping;
            (this.AssociatedObject.View as TableView).CellValueChanged += GridCellValueChanged;
            (this.AssociatedObject.View as TableView).CellValueChanging += CellValueChanging;
        }

        void AssociatedObject_EndGrouping(object sender, RoutedEventArgs e) {
            if(this.AssociatedObject.GroupCount == 0) { return; }

            var res = this.GetAllRowHandles();
            List<int> parents = new List<int>();
            int j = res.Count;
            while(--j >= 0) {
                var parent = this.AssociatedObject.GetParentRowHandle(res[j]);
                if(!parents.Contains(parent)) {
                    parents.Add(parent);
                    do {
                        parent = this.AssociatedObject.GetParentRowHandle(parent);
                        if(!parents.Contains(parent) && this.AssociatedObject.IsValidRowHandle(parent)) {
                            parents.Add(parent);
                        }

                    } while(this.AssociatedObject.IsValidRowHandle(parent));
                }
            }

            int g = this.AssociatedObject.GroupCount;
            while(--g >= 0) {
                var levelItems = parents.Where(c => this.AssociatedObject.GetRowLevelByRowHandle(c) == g);
                if(levelItems == null) { continue; }
                foreach(var item in levelItems) {
                    var b = this.GetRowChildrenState(item);
                    var dic = (Dictionary<int, bool?>)this.AssociatedObject.GetRowState(item).GetValue(CheckableGroupBehavior.CheckListProperty);
                    dic[g] = b;
                    this._LocalSet = true;
                    this.AssociatedObject.GetRowState(item).SetValue(CheckableGroupBehavior.CheckListProperty, new Dictionary<int, bool?>(dic));
                    this._LocalSet = false;
                }
            }
        }

        protected List<int> GetAllRowHandles() {
            List<int> result = new List<int>();
            foreach(var item in this.AssociatedObject.ItemsSource as IEnumerable<object>) {
                int handle = this.AssociatedObject.FindRowByValue("", item);
                result.Add(handle);
            }

            return result;
        }

        protected void CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) {
            switch(e.Action) {
                case System.Collections.Specialized.NotifyCollectionChangedAction.Add:
                case System.Collections.Specialized.NotifyCollectionChangedAction.Replace:
                    foreach(var item in e.NewItems) {
                        this.AttachItem(item);
                    }
                    break;
                default:
                    break;
            }
        }

        public void PropertyChanged(object obj, PropertyChangedEventArgs args) {
            if(!this._LocalSet && args.PropertyName == this.CheckableProperty) {
                var grid = this.AssociatedObject;
                int rowHandle = grid.FindRowByValue("", obj);
                this.WorkOutCell(rowHandle);
            }
        }

        protected void CellValueChanging(object sender, CellValueChangedEventArgs e) {
            var grid = this.AssociatedObject;
            grid.View.PostEditor();
        }

        protected void GridCellValueChanged(object sender, CellValueChangedEventArgs e) {
            if(e.Column.FieldName == this.CheckableProperty && !this._LocalSet) {
                this._LocalSet = true;
                this.WorkOutCell(e.RowHandle);
                this._LocalSet = false;
            }
        }

        protected void WorkOutCell(int rowHandle) {
            var grid = this.AssociatedObject;
            int groupRowHandle = grid.GetParentRowHandle(rowHandle);
            if(grid.IsValidRowHandle(groupRowHandle) && grid.IsGroupRowHandle(groupRowHandle)) {
                bool? value = this.GetRowChildrenState(groupRowHandle);
                var dic = new Dictionary<int, bool?>(this.GetUpdatedDic(groupRowHandle, value));
                grid.GetRowState(groupRowHandle).SetCurrentValue(CheckableGroupBehavior.CheckListProperty, dic);
                this.WorkOutCell(groupRowHandle);
            }
        }

        protected bool? GetRowChildrenState(int rowHandle) {
            var grid = this.AssociatedObject;
            bool? result = null, b;
            if(grid.IsValidRowHandle(rowHandle) && grid.IsGroupRowHandle(rowHandle)) {
                int count = grid.GetChildRowCount(rowHandle);
                int i = -1;
                while(++i < count) {
                    var childHandle = grid.GetChildRowHandle(rowHandle, i);
                    int childLevel = grid.GetRowLevelByRowHandle(childHandle);
                    if(grid.IsGroupRowHandle(childHandle)) {
                        var dic = (Dictionary<int, bool?>)grid.GetRowState(childHandle).GetValue(CheckableGroupBehavior.CheckListProperty);
                        b = dic[childLevel];
                    } else {
                        var column = grid.Columns.Where(c => c.FieldName == this.CheckableProperty).FirstOrDefault();
                        if(column != null) {
                            b = (bool)grid.GetCellValue(childHandle, column);
                        } else {
                            b = false;
                        }
                    }

                    if(b.HasValue) {
                        if(result.HasValue) {
                            if(result.Value ^ b.Value) {
                                return null;
                            } 
                        } else {
                            result = b;
                        }
                    } else {
                        return null;
                    }
                }
            }

            return result;
        }

        protected Dictionary<int, bool?> GetUpdatedDic(int rowHandle, bool? value) {
            Dictionary<int, bool?> result = null;
            if(this.AssociatedObject.IsValidRowHandle(rowHandle)){
                int groupLevel = this.AssociatedObject.GetRowLevelByRowHandle(rowHandle);
                var rowState = this.AssociatedObject.GetRowState(rowHandle);
                result = (Dictionary<int, bool?>)rowState.GetValue(CheckableGroupBehavior.CheckListProperty);
                result[groupLevel] = value;
            }

            return result;
        }

        protected void WorkOutGroup(int rowHandle, bool? value) {
            var grid = this.AssociatedObject;
            var parentRowState = grid.GetRowState(rowHandle);
            if(grid.IsValidRowHandle(rowHandle) && grid.IsGroupRowHandle(rowHandle)) {
                int count = grid.GetChildRowCount(rowHandle);
                int i = -1;
                while(++i < count) {
                    var childHandle = grid.GetChildRowHandle(rowHandle, i);
                    if(grid.IsGroupRowHandle(childHandle)) {
                        var rowState = this.AssociatedObject.GetRowState(childHandle);
                        rowState.SetCurrentValue(CheckableGroupBehavior.CheckListProperty, new Dictionary<int, bool?>(this.GetUpdatedDic(childHandle, value)));
                        this.WorkOutGroup(childHandle, value);
                    } else {
                        var column = grid.Columns.Where(c => c.FieldName == this.CheckableProperty).FirstOrDefault();
                        if(column != null) {
                            grid.SetCellValue(childHandle, column, value);
                        }
                    }
                }

                this.WorkOutCell(rowHandle);
            }
        }

        #region Template For GroupRowValue

        protected DataTemplate GetGroupValueTemplate() {
            string xamlDataTemplate = "<DataTemplate>";
            xamlDataTemplate += "<self:GroupCheckBox Content=\"{Binding Value}\" Margin=\"8,0,0,0\"  CheckStates=\"{Binding Path=RowData.RowState.(self:CheckableGroupBehavior.CheckList), Mode=TwoWay}\"/>";
            xamlDataTemplate += "</DataTemplate>";
            ParserContext context = new ParserContext();
            context.XmlnsDictionary.Add("", "http://schemas.microsoft.com/winfx/2006/xaml/presentation");
            context.XmlnsDictionary.Add("x", "http://schemas.microsoft.com/winfx/2006/xaml");
            context.XmlnsDictionary.Add("self", "clr-namespace:DevExpress.Example03;assembly=DevExpress.Example03");
            DataTemplate template = (DataTemplate)XamlReader.Parse(xamlDataTemplate, context);
            return template;
        }

        #endregion Template for GroupRowValue
    }
}
